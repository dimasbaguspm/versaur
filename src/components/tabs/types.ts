/**
 * Tabs types for Versaur UI
 */
import type { ReactNode, HTMLAttributes } from 'react'

/**
 * TabsContextValue: Shared state for compound/context Tabs
 */
export interface TabsContextValue {
  /** Currently active tab value */
  activeTab: string
  /** Set active tab */
  setActiveTab: (tab: string) => void
  /** Default color for triggers */
  color?: 'primary' | 'secondary' | 'tertiary' | 'ghost' | 'neutral'
}

/**
 * TabsRootProps: Props for Tabs root
 */
export interface TabsRootProps extends HTMLAttributes<HTMLDivElement> {
  /** Controlled tab value */
  value: string
  /** Callback when tab changes */
  onValueChange: (tab: string) => void
  /** Color for triggers */
  color?: 'primary' | 'secondary' | 'tertiary' | 'ghost' | 'neutral'
  /** Children: Tabs.Trigger */
  children: ReactNode
}

/**
 * TabsTriggerProps: Individual tab button
 */
export interface TabsTriggerProps extends HTMLAttributes<HTMLButtonElement> {
  /** Tab value (unique string) */
  value: string
  /** Children: tab label */
  children: ReactNode
}

/**
 * TabsIndicatorProps: Animated underline indicator for active tab
 */
export interface TabsIndicatorProps extends HTMLAttributes<HTMLDivElement> {
  left: number
  width: number
}
