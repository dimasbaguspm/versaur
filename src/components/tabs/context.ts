// Tabs context for compound/context pattern
import { createContext, useContext } from 'react'
import type { TabsContextValue } from './types'

export const TabsContext = createContext<TabsContextValue | null>(null)

export const useTabsContext = () => {
  const context = useContext(TabsContext)
  if (!context) {
    throw new Error('useTabsContext must be used within a TabsProvider')
  }
  return context
}
