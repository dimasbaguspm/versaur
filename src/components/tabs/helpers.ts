// Tabs helpers: variants, sizes, cva
import { cva } from 'class-variance-authority'

export const tabsTriggerVariants = cva(
  'inline-flex items-center justify-center transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none select-none px-4 py-2 text-base bg-transparent relative border-b-2 border-transparent',
  {
    variants: {
      active: {
        true: 'font-medium',
        false: 'font-normal',
      },
      color: {
        primary: '',
        secondary: '',
        tertiary: '',
        ghost: '',
        neutral: '',
      },
    },
    compoundVariants: [
      // Primary
      {
        color: 'primary',
        active: true,
        class: 'text-coral border-coral',
      },
      {
        color: 'primary',
        active: false,
        class: 'text-foreground hover:text-coral',
      },
      // Secondary
      {
        color: 'secondary',
        active: true,
        class: 'text-sage border-sage',
      },
      {
        color: 'secondary',
        active: false,
        class: 'text-foreground hover:text-sage',
      },
      // Tertiary
      {
        color: 'tertiary',
        active: true,
        class: 'text-mist border-mist',
      },
      {
        color: 'tertiary',
        active: false,
        class: 'text-foreground hover:text-mist',
      },
      // Ghost
      {
        color: 'ghost',
        active: true,
        class: 'text-slate border-slate',
      },
      {
        color: 'ghost',
        active: false,
        class: 'text-foreground hover:text-slate',
      },
      // Neutral
      {
        color: 'neutral',
        active: true,
        class: 'text-foreground border-light-gray',
      },
      {
        color: 'neutral',
        active: false,
        class: 'text-foreground hover:text-light-gray',
      },
    ],
    defaultVariants: {
      active: false,
      color: 'primary',
    },
  }
)

export const tabsContainerVariants = cva(
  'flex flex-row w-full border-b border-neutral overflow-x-auto whitespace-nowrap',
  {
    variants: {},
    defaultVariants: {},
  }
)

export const tabsIndicatorVariants = cva(
  'absolute bottom-0 h-0.5 rounded transition-all duration-300',
  {
    variants: {
      color: {
        primary: 'bg-coral',
        secondary: 'bg-sage',
        tertiary: 'bg-mist',
        ghost: 'bg-slate',
        neutral: 'bg-light-gray',
      },
    },
    defaultVariants: {
      color: 'primary',
    },
  }
)
