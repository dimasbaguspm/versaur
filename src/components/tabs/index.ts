export { Tabs } from './tabs'
export * from './types'
