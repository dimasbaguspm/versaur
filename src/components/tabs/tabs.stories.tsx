/**
 * Tabs component stories for Versaur UI
 * Follows Material Design guidelines and Versaur codebase standards
 * Demonstrates vertical orientation, color variants, and accessibility
 * Includes interactive controlled example
 */
import { useState } from 'react'
import type { Meta, StoryObj } from '@storybook/react'
import { Tabs } from './tabs'

const meta: Meta<typeof Tabs> = {
  title: 'Navigation/Tabs',
  component: Tabs,
  tags: ['autodocs'],
}
export default meta

type Story = StoryObj<typeof Tabs>

/**
 * Example: Basic vertical tabs, content rendered externally
 */
function BasicExample() {
  const [tab, setTab] = useState('tab1')
  return (
    <>
      <Tabs value={tab} onValueChange={setTab}>
        <Tabs.Trigger value='tab1'>Tab 1</Tabs.Trigger>
        <Tabs.Trigger value='tab2'>Tab 2</Tabs.Trigger>
        <Tabs.Trigger value='tab3'>Tab 3</Tabs.Trigger>
      </Tabs>
      <div className='mt-4'>
        {tab === 'tab1' && <div>Content for Tab 1</div>}
        {tab === 'tab2' && <div>Content for Tab 2</div>}
        {tab === 'tab3' && <div>Content for Tab 3</div>}
      </div>
    </>
  )
}

/**
 * Basic vertical tabs
 */
export const Basic: Story = {
  render: () => <BasicExample />,
}

/**
 * Example: Tabs with all color variants
 */
function ColorVariantsExample() {
  const [tab, setTab] = useState('primary')
  const colors: Array<{
    key: 'primary' | 'secondary' | 'tertiary' | 'ghost' | 'neutral'
    label: string
  }> = [
    { key: 'primary', label: 'Coral' },
    { key: 'secondary', label: 'Sage' },
    { key: 'tertiary', label: 'Mist' },
    { key: 'ghost', label: 'Slate' },
    { key: 'neutral', label: 'Light Gray' },
  ]
  return (
    <div className='space-y-6'>
      {colors.map(({ key, label }) => (
        <div key={key}>
          <div className='mb-2 font-semibold text-sm'>{label}</div>
          <Tabs value={tab} onValueChange={setTab} color={key}>
            <Tabs.Trigger value='primary'>Primary</Tabs.Trigger>
            <Tabs.Trigger value='secondary'>Secondary</Tabs.Trigger>
            <Tabs.Trigger value='tertiary'>Tertiary</Tabs.Trigger>
            <Tabs.Trigger value='ghost'>Ghost</Tabs.Trigger>
            <Tabs.Trigger value='neutral'>Neutral</Tabs.Trigger>
          </Tabs>
        </div>
      ))}
    </div>
  )
}

/**
 * Tabs with color variants
 */
export const ColorVariants: Story = {
  render: () => <ColorVariantsExample />,
}

/**
 * Example: Tabs with scroll behavior and many items
 */
function ScrollBehaviorExample() {
  const [tab, setTab] = useState('tab5')
  const tabItems = Array.from({ length: 15 }, (_, i) => `tab${i + 1}`)
  return (
    <div className='max-w-xl border p-4'>
      <Tabs value={tab} onValueChange={setTab}>
        {tabItems.map(key => (
          <Tabs.Trigger key={key} value={key}>
            {`Tab ${key.replace('tab', '')}`}
          </Tabs.Trigger>
        ))}
      </Tabs>
      <div className='mt-4'>{`Content for ${tab}`}</div>
    </div>
  )
}

/**
 * Tabs with scroll behavior
 */
export const ScrollBehavior: Story = {
  render: () => <ScrollBehaviorExample />,
}
