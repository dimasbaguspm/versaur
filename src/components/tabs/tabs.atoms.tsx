// Tabs atoms for Versaur UI
import React from 'react'
import { useTabsContext } from './context'
import type { TabsTriggerProps, TabsIndicatorProps } from './types'
import { cn } from '@/utils/cn'
import { tabsTriggerVariants, tabsIndicatorVariants } from './helpers'

/**
 * TabsTrigger: Individual tab button
 */
export const TabsTrigger: React.FC<TabsTriggerProps> = ({
  value,
  children,
  className,
  ...props
}) => {
  const { activeTab, setActiveTab, color } = useTabsContext()
  const isActive = activeTab === value

  return (
    <button
      role='tab'
      id={`tabs-trigger-${value}`}
      aria-selected={isActive}
      aria-controls={`tabs-content-${value}`}
      tabIndex={isActive ? 0 : -1}
      className={cn(
        tabsTriggerVariants({
          active: isActive,
          color,
        }),
        className
      )}
      onClick={() => setActiveTab(value)}
      type='button'
      {...props}
    >
      {children}
    </button>
  )
}

/**
 * TabsIndicator: Animated underline indicator for active tab
 */
export const TabsIndicator: React.FC<TabsIndicatorProps> = ({
  left,
  width,
  className,
  ...props
}) => {
  const { color } = useTabsContext()
  return (
    <div
      aria-hidden
      className={cn(tabsIndicatorVariants({ color }), className)}
      style={{
        left,
        width,
        pointerEvents: 'none',
      }}
      {...props}
    />
  )
}
