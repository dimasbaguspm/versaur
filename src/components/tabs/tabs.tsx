// Compound/context Tabs component for Versaur UI
// See context.ts for shared state
import React, { useRef } from 'react'
import { TabsContext } from './context'
import type { TabsRootProps } from './types'
import { TabsTrigger, TabsIndicator } from './tabs.atoms'
import { cn } from '@/utils/cn'
import { tabsContainerVariants } from './helpers'
import { useTabIndicatorAndFocus } from './use-tab-indicator'

/**
 * TabsRoot: Compound/context root for Tabs
 * @see https://m3.material.io/components/tabs/guidelines
 */
const TabsRoot: React.FC<TabsRootProps> = ({
  value,
  onValueChange,
  children,
  className,
  color = 'primary',
  ...props
}) => {
  const contextValue = {
    activeTab: value,
    setActiveTab: onValueChange,
    color,
  }

  const containerRef = useRef<HTMLDivElement>(null)
  const indicatorStyle = useTabIndicatorAndFocus(value, containerRef, children)

  return (
    <TabsContext.Provider value={contextValue}>
      <div
        role='tablist'
        className={cn(tabsContainerVariants(), className)}
        ref={containerRef}
        style={{ position: 'relative' }}
        {...props}
      >
        {children}
        <TabsIndicator
          left={indicatorStyle.left}
          width={indicatorStyle.width}
        />
      </div>
    </TabsContext.Provider>
  )
}

export const Tabs = Object.assign(TabsRoot, {
  Trigger: TabsTrigger,
})
