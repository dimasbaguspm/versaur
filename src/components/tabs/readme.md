# Tabs

Compound/context Tabs component for Versaur UI. Follows Material Design guidelines and Versaur
codebase standards.

## Features

- Compound/context pattern: Tabs.Trigger
- Strictly typed props and context
- ARIA roles and attributes for accessibility
- Mobile-first, responsive, and themeable
- Supports horizontal and vertical orientation
- Color and size variants via Versaur system
- Scroll behavior for large tab lists

## Usage

```tsx
import { Tabs } from '@/components/tabs'

// Basic controlled tabs
function BasicExample() {
  const [tab, setTab] = useState('tab1')
  return (
    <>
      <Tabs value={tab} onValueChange={setTab}>
        <Tabs.Trigger value='tab1'>Tab 1</Tabs.Trigger>
        <Tabs.Trigger value='tab2'>Tab 2</Tabs.Trigger>
        <Tabs.Trigger value='tab3'>Tab 3</Tabs.Trigger>
      </Tabs>
      <div className='mt-4'>
        {tab === 'tab1' && <div>Content for Tab 1</div>}
        {tab === 'tab2' && <div>Content for Tab 2</div>}
        {tab === 'tab3' && <div>Content for Tab 3</div>}
      </div>
    </>
  )
}

// Color variants
function ColorVariantsExample() {
  const [tab, setTab] = useState('primary')
  return (
    <Tabs value={tab} onValueChange={setTab} color={tab}>
      <Tabs.Trigger value='primary'>Primary</Tabs.Trigger>
      <Tabs.Trigger value='secondary'>Secondary</Tabs.Trigger>
      <Tabs.Trigger value='tertiary'>Tertiary</Tabs.Trigger>
      <Tabs.Trigger value='ghost'>Ghost</Tabs.Trigger>
      <Tabs.Trigger value='neutral'>Neutral</Tabs.Trigger>
    </Tabs>
  )
}

// Scroll behavior for many tabs
function ScrollBehaviorExample() {
  const [tab, setTab] = useState('tab5')
  const tabItems = Array.from({ length: 15 }, (_, i) => `tab${i + 1}`)
  return (
    <div className='max-w-xl border p-4'>
      <Tabs value={tab} onValueChange={setTab}>
        {tabItems.map(key => (
          <Tabs.Trigger key={key} value={key}>
            {`Tab ${key.replace('tab', '')}`}
          </Tabs.Trigger>
        ))}
      </Tabs>
      <div className='mt-4'>{`Content for ${tab}`}</div>
    </div>
  )
}
```

## Props

See `types.ts` for full prop documentation.
