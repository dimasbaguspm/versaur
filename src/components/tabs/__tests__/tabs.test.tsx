import { render, screen, fireEvent } from '@testing-library/react'
import { Tabs } from '@/components/tabs'
import { useState } from 'react'

describe('Tabs', () => {
  it('renders tab triggers and content (controlled)', () => {
    function TestTabs() {
      const [tab, setTab] = useState('tab1')
      return (
        <>
          <Tabs value={tab} onValueChange={setTab}>
            <Tabs.Trigger value='tab1'>Tab 1</Tabs.Trigger>
            <Tabs.Trigger value='tab2'>Tab 2</Tabs.Trigger>
            <Tabs.Trigger value='tab3'>Tab 3</Tabs.Trigger>
          </Tabs>
          <div data-testid='content'>
            {tab === 'tab1' && <div>Content for Tab 1</div>}
            {tab === 'tab2' && <div>Content for Tab 2</div>}
            {tab === 'tab3' && <div>Content for Tab 3</div>}
          </div>
        </>
      )
    }
    render(<TestTabs />)
    expect(screen.getByRole('tab', { name: 'Tab 1' })).toBeInTheDocument()
    expect(screen.getByRole('tab', { name: 'Tab 2' })).toBeInTheDocument()
    expect(screen.getByRole('tab', { name: 'Tab 3' })).toBeInTheDocument()
    expect(screen.getByTestId('content')).toHaveTextContent('Content for Tab 1')
  })

  it('switches tab on trigger click (controlled)', () => {
    function TestTabs() {
      const [tab, setTab] = useState('tab1')
      return (
        <>
          <Tabs value={tab} onValueChange={setTab}>
            <Tabs.Trigger value='tab1'>Tab 1</Tabs.Trigger>
            <Tabs.Trigger value='tab2'>Tab 2</Tabs.Trigger>
          </Tabs>
          <div data-testid='content'>
            {tab === 'tab1' && <div>Content for Tab 1</div>}
            {tab === 'tab2' && <div>Content for Tab 2</div>}
          </div>
        </>
      )
    }
    render(<TestTabs />)
    fireEvent.click(screen.getByRole('tab', { name: 'Tab 2' }))
    expect(screen.getByTestId('content')).toHaveTextContent('Content for Tab 2')
  })

  it('matches snapshot', () => {
    function TestTabs() {
      const [tab, setTab] = useState('tab1')
      return (
        <Tabs value={tab} onValueChange={setTab}>
          <Tabs.Trigger value='tab1'>Tab 1</Tabs.Trigger>
          <Tabs.Trigger value='tab2'>Tab 2</Tabs.Trigger>
        </Tabs>
      )
    }
    const { asFragment } = render(<TestTabs />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders color variants', () => {
    function ColorTabs() {
      const [tab, setTab] = useState('primary')
      return (
        <Tabs value={tab} onValueChange={setTab} color={tab as 'primary'}>
          <Tabs.Trigger value='primary'>Primary</Tabs.Trigger>
          <Tabs.Trigger value='secondary'>Secondary</Tabs.Trigger>
          <Tabs.Trigger value='tertiary'>Tertiary</Tabs.Trigger>
          <Tabs.Trigger value='ghost'>Ghost</Tabs.Trigger>
          <Tabs.Trigger value='neutral'>Neutral</Tabs.Trigger>
        </Tabs>
      )
    }
    render(<ColorTabs />)
    expect(screen.getByRole('tab', { name: 'Primary' })).toBeInTheDocument()
    expect(screen.getByRole('tab', { name: 'Secondary' })).toBeInTheDocument()
    expect(screen.getByRole('tab', { name: 'Tertiary' })).toBeInTheDocument()
    expect(screen.getByRole('tab', { name: 'Ghost' })).toBeInTheDocument()
    expect(screen.getByRole('tab', { name: 'Neutral' })).toBeInTheDocument()
  })

  it('handles scroll behavior for many tabs', () => {
    function ScrollTabs() {
      const [tab, setTab] = useState('tab5')
      const tabItems = Array.from({ length: 15 }, (_, i) => `tab${i + 1}`)
      return (
        <Tabs value={tab} onValueChange={setTab}>
          {tabItems.map(key => (
            <Tabs.Trigger key={key} value={key}>
              {`Tab ${key.replace('tab', '')}`}
            </Tabs.Trigger>
          ))}
        </Tabs>
      )
    }
    render(<ScrollTabs />)
    expect(screen.getByRole('tab', { name: 'Tab 5' })).toBeInTheDocument()
    fireEvent.click(screen.getByRole('tab', { name: 'Tab 10' }))
    expect(screen.getByRole('tab', { name: 'Tab 10' })).toHaveAttribute(
      'aria-selected',
      'true'
    )
  })
})
