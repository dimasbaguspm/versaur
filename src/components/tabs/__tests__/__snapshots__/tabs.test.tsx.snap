// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tabs > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-row w-full border-b border-neutral overflow-x-auto whitespace-nowrap"
    role="tablist"
    style="position: relative;"
  >
    <button
      aria-controls="tabs-content-tab1"
      aria-selected="true"
      class="inline-flex items-center justify-center transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none select-none px-4 py-2 text-base bg-transparent relative border-b-2 font-medium text-coral border-coral"
      id="tabs-trigger-tab1"
      role="tab"
      tabindex="0"
      type="button"
    >
      Tab 1
    </button>
    <button
      aria-controls="tabs-content-tab2"
      aria-selected="false"
      class="inline-flex items-center justify-center transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none select-none px-4 py-2 text-base bg-transparent relative border-b-2 border-transparent font-normal text-foreground hover:text-coral"
      id="tabs-trigger-tab2"
      role="tab"
      tabindex="-1"
      type="button"
    >
      Tab 2
    </button>
    <div
      aria-hidden="true"
      class="absolute bottom-0 h-0.5 rounded transition-all duration-300 bg-coral"
      style="left: 0px; width: 0px; pointer-events: none;"
    />
  </div>
</DocumentFragment>
`;
